### ISSUE TITLE
Proper way of send error handling

### ISSUE THREAD

**Issue by maciej-miszczak** [TakahikoKawasaki/nv-websocket-client#44]
Hello,

it's not clear for me how to handle send errors correctly. In my Service I create an AsyncTask, which handles ArrayBlockingQueue in a loop. Then I run connectAsynchronously() in this service. Into the mentioned queue I add messages I want to send. As far as I understand, sendText() returns even in case of an error - the onSendError() callback is then called, but all the code after sendText() invocation is executed. How can I then implement some way to buffer my data in a database in case of transmission problems? I tried to save the data in a database in onSendError() or onFrameUnsent() and clear the queue head in onFrameSent(), but it appears, that onFrameSent() is executed also after a ping frame, so I would probably have to check the queue head for the getPayloadText() of a sent frame. Am I right?

Cheers,
Matthew

**Comment by TakahikoKawasaki** [comment]
There will be several approaches to achieve your goal. The approach you should take will vary depending on the level of assurance you want. For example, if you want to make sure that frames have reached your server successfully, your server should return a response to indicate it and your client should recognize it.

BTW, `WebSocketFrame` has methods to indicate its frame type as listed below.

- `int getOpcode()`
- `boolean isBinaryFrame()`
- `boolean isCloseFrame()`
- `boolean isContinuationFrame()`
- `boolean isControlFrame()`
- `boolean isDataFrame()`
- `boolean isPingFrame()`
- `boolean isPongFrame()`
- `boolean isTextFrame()`

You can call these methods in callback methods that take a `WebSocketFrame` instance as an argument. For example, `onFrameSent()`.

**Comment by maciej-miszczak** [comment]
Thanks for pointing the necessity of some kind of acknowledgements from server. I will take it into consideration. Unfortunately now I noticed another problem - with keeping the connection alive in case of falling into a sleep state. Now trying to perform a partial wake and send a ping frame ex. once a minute.

**Comment by TakahikoKawasaki** [comment]
FYI: Once you call `WebSocket.setPingInterval(long interval)`, ping frames are sent periodically.

**Comment by maciej-miszczak** [comment]
Correct, but if I lock my phone and its display turns off, after some time it stops to send the ping frames, most probably due to the sleep mode.

**Comment by eku** [comment]
Scheduling an Android wakeup broadcast is up to you and far out of the scope of a websocket library.

**Comment by BoD** [comment]
Hi, maybe I'm late to the party, but I've just started to use this library and so far it looks very solid :)

However I too, would like to see some kind of mechanism such as a "blocking mode", meaning: sendXxx methods that try to send the data immediately (blocking) and throw an Exception if they can't. This would simplify error handling code IMHO - and behave more like the "plain" socket Java API.

I am currently implementing an "event reporting" system, and I need to know if one event has been sent or not (to delete events from the local storage after they were successfully sent).  To do that with the current API, I will need to use the WebSocketListener interface and keep track of what has been sent there. A bit convoluted.

Thoughts?

**Comment by eku** [comment]
@BoD What about the callbacks ```onSendingFrame``` and ```onFrameSent```?

**Comment by BoD** [comment]
@eku Well, that's what I meant by "I will need to use the WebSocketListener interface".

But to do this is not trivial, I will need to implement some kind of asynchronous mechanism: one thread sends the data, another thread is notified that the data was successfully sent or not, and then has to notify the first thread somehow.

My point was that it would be a lot easier (and more similar to the plain socket API), to just have sendXxx methods be blocking and throw an Exception if it didn't work.

**Comment by eku** [comment]
@BoD this websocket implementation works asynchroniously. Blocking network calls are a bad way.

**Comment by BoD** [comment]
@eku Hmmm I don't know, for instante to do what I'm trying to do, blocking would not be "bad" ;)

In any case I was suggesting a second API to supplement the current one - not to replace it.

Some popular network libraries do this for instance in "Retrofit" you have blocking calls, and also async ones that return a Future, for instance.
